--- 
title: "tanggle: An R package for the visualization of implicit and explicit phylogenetic networks"
author: klaus, marta, leann, francisco, eran, claudia
site: bookdown::bookdown_site
output: 
   pdf_document:
    toc: false
    citation_package: natbib
    keep_tex: yes
header-includes:
- \usepackage[dvipsnames]{xcolor}
- \newcommand{\missing}[1]{\textcolor{red}{\textbf{#1}}}
- \newcommand{\revision}[1]{\textcolor{blue}{#1}}
- \usepackage{array,ragged2e}
- \usepackage{float}
documentclass: IEEEtran
bibliography: ggnetworx.bib
biblio-style: apalike
geometry: margin=0.8in
link-citations: yes
---


```{r set-options, echo=FALSE, include=FALSE}
knitr::opts_chunk$set(message=FALSE, warning=FALSE, 
                      error=FALSE, echo=FALSE)

```


\begin{center}
\textbf{Abstract} 
\end{center}

Here we present an extension to the widely used visualization package `ggtree` to the case of phylogenetic networks. Our packages allows a variety of input data from DNA sequences to extended Newick format, and allows the user to plot the networks with great flexibility given that the many functionalities such as...

# Introduction

Phylogenetic trees are used to represent evolutionary relationships between organisms. However, the tree is a strictly bifurcating structure that does not capture reticulate evolution like horizontal gene transfer, hybridization or introgression.

Recent years have seen an explosion of phylogenetic network inference methods (\missing{cite everybody}). These methods can be divided into two main classes: methods to reconstruct implicit (or split) networks, and methods to reconstruct explicit networks. Implicit networks are data-displayed objects that include two (or more) options for every uncertain split. For example, in figure \missing{bla}, there are two possible ways to resolve the uncertain split, but the input data (DNA sequences or gene trees) does not provide enough information to distinguish which one is the "correct" split. Or alternatively, both splits are "correct" in that there is a pattern of gene flow among these species. Split networks are fast to reconstruct, but they are limited in their interpretations are they cannot tell us if the discordant splits are due to gene tree estimation error, ILS or actual gene flow.
Explicit networks, on the contrary, assign a biological mechanism to every internal node, and thus they are easier to interpret. For example, in figure \missing{bla}, the hybrid node corresponds to some reticulation event, and the tree nodes correspond to speciation events. Explicit networks are still limited in that they cannot identify the specific biological mechanism behind a hybrid node (hybridization, horizontal gene transfer, introgression), but hybrid edges are parametrized by an inheritance probability ($\gamma$) that can provide more information. For example, an estimated $\hat{\gamma} \approx 0.5$ could indicate a hybridization event, whereas smaller estimated $\hat{\gamma} \approx 0.01$ could refer to horizontal gene transfer. The downside of explicit networks is that they are harder to reconstruct, and existing inference methods\missing{(reference)} are not scalable enough to handle big data.

\vspace{-0.25cm}
## Extended Newick format
Traditionally, phylogenetic trees are written in Newick format (\missing{reference}). To represent network objects, we use the
extended Newick format (Morin and Moret 2006; Than et al.
2008; Cardona et al. 2008a, 2008b). 
This format uses the concept of minor
hybrid edges (edges with $\gamma < 0.5$) and major hybrid edges
($\gamma > 0.5$). By default, we detach the minor hybrid edge at each hybrid node to write the extended Newick
description of a network as we would for a tree, with a repeated label, that of the hybrid node ('#H1' in figure \@ref(fig:netNewick)). This description can include edge
information, formatted as :length:support:$\gamma$.

```{r netNewick, out.width='35%', fig.cap="Extended Newick format for networks: (((A,(B)\\#H1),(C,\\#H1)),D); The network is written as a tree with two nodes having the same label \\#H1.",fig.show='hold', fig.align='center'}
knitr::include_graphics(c('figures/extendedNewick.pdf', 'figures/extendedNewick2.pdf'))
```

For example, the parenthetical format of the network in figure \@ref(fig:netNewick) can include $\gamma$ values: 

`(((A,(B)#H1:::0.8),(C,#H1:::0.2)),D);`, 

which are written after three colons, because there is no information about branch length nor support for the hybrid edges. Other internal edges (tree edges) have information about branch lengths, which follow just one colon. These tree edges do not have information about $\gamma$, because all tree edges have $\gamma=1$.

# `tanggle` package

The `tanggle` package extends the already widely used `ggtree` (\missing{references}) to implicit and explicit phylogenetic networks. Our package has two main functions: `ggsplitnet` that plots split networks, and `ggevonet` that plots explicit networks.

\vspace{-0.25cm}
## Reconstructing and plotting a split network

There are three types of input data that our package takes to get a split network:

1. Nexus file created with SplitsTree (\missing{reference}) which is read with the `read.nexus.networx` function in the `phangorn` package.
2. Collection of gene trees (e.g. from RAxML\missing{(reference)} or MrBayes\missing{(reference)}) in a nexus file, or in a text file with one row per gene tree in Newick format. These trees are read with the functions `read.tree` or `read.nexus`. A consensus split network is then computed using the `consensusNet` function in the `phangorn` package \missing{(reference)}. This function implements the algorithm in \missing{(reference)}.
3. Sequences in nexus, fasta or phylip format read with the `read.phyDat` function in `phangorn` package or the `read.dna` function in `ape` package. Distance matrices are computed then for specific models of evolution using the function `dist.ml` from `phangorn` package or the function `dist.dna` from `ape` package. From the distance matrix, a split network is reconstructed using the `neighborNet` function in the `phangorn` package. This function implements the algorithm in \missing{(reference)}. In this case, there is the option to estimate branch lengths as well (which is not usually the case for split networks) with the function `splitsNetworks` in the `phangorn` package.

After any of these three steps, we have a `networkx` object that represents a split network by two matrices: 1) standard edge matrix with two columns for the two nodes connecting every edge (row), and 2) split vector with integer identifier for every edge. Each entry in this vector represents the split that each edge is representing. In split networks, there are more edges than splits as two (or more) edges could represent the same split (see figure \missing{bla}).

The `ggsplitnet` function takes a `networkx` object as input, and produces a plot. 

```{r, echo=TRUE}
library(ggnetworx)
data(yeast, package="phangorn")
dm <- phangorn::dist.ml(yeast) 
nnet <- phangorn::neighborNet(dm)
p = ggsplitnet(nnet) + geom_tiplab2(cex=4.0)
```

```{r nnet, out.width='60%', fig.cap="Split network for yeast data", fig.pos='H', fig.align='center'}
p
```



## Plotting explicit networks

Algorithms to reconstruct explicit networks are more computationally intensive than those to reconstruct split networks. Thus, to the best of our knowledge, there are no R functions that would estimate an explicit network from sequences or from gene trees. To obtain an explicit network, users need to use existing tools like PhyloNet\missing{(reference)}, BEAST\missing{(reference)} or SNaQ\missing{(reference)}.
These methods take sequences or gene trees as input and estimate an explicit network which is represented in extended Newick format. The explicit network is read with the `ape` function `read.evonet` and an `evonet` object is created. 

The `evonet` object has two matrices: 1) standard edge matrix with two columns for the two nodes connecting every edge (row), and 2) hybrid edges matrix with two columns for the nodes connected by the hybrid edge (row).

```{r, echo=TRUE}
library(ape)
net = "((a:2,(b:1)#H1:1):1,(#H1,c:1):2);"
enet <- read.evonet(text=net)
p2 = ggevonet(enet) + geom_tiplab() 
```

```{r enet, out.width='60%', fig.cap="Explicit network from extended Newick format", fig.pos='H', fig.align='center'}
p2
```


# Split network for the _Neobatrachus_ genus 

\missing{(Marta: add here short description of genus)}.

```{r, echo=FALSE}
library(phangorn)
trees = read.nexus("data-analysis/output.nex")
frognet = consensusNet(trees)
frog.fortify <- ggnetworx:::fortify.networx(frognet)
branch <- c(1:18) ##this is not what branch should be
frog.fortify <- data.frame(frog.fortify, branch)
p3 = ggsplitnet(frognet)+geom_tiplab2()+
    geom_hilight_encircle(node=15, alpha=0.4, fill = "steelblue",
                          branch.length =frog.fortify$branch.length,
                          extend = 0) +
    geom_hilight_encircle(node=16, alpha=0.4, fill = "green",
                          branch.length = frog.fortify$branch.length,
                          extend = 0)
```

```{r frognet, out.width='60%', fig.cap="Explicit network for Neobatrachus genus frogs", fig.pos='H', fig.align='center'}
p3
```

# Explicit network for whales

Baleen whales (Mysticeti) represent the largest animals on Earth. Their evolutionary history is complex given the lack of physical barriers in the ocean, and the fact that they hybridize frequently.
Arnason et al \missing{(cite Arnason 2018)} reconstruct the network-like evolutionary history from a variety of methods, using 34,192 genome fragments, each with 20 kbp long. Reconstructed GF trees with RAxML and GTR with gamma-distributed rate variation with invariable sites. Here we use SNaQ \missing{(cite)} to reconstruct the network with two hybridization events, one of which was already reported in the original paper (from blue whale to gray whale). SNaQ produces the network in extended newick format, which we can read into R and then plot with `tanggle`.

```{r, echo=FALSE}
whalesnet = read.evonet(text="(((bowhead,north-atlantic):9.055782863234407)#H13:2.6340479153169807::0.8331903627712247,(minke,((sei:9.41949690722463,(blue,#H:0.0::0.013010091110208823)):1.6484045144986545,((#H13:0.19346605673125924::0.16680963722877526,(gray:1.7595289284120899)#H:8.238071084864815::0.9869899088897912):0.35164828246609625,(humpback,fin):2.146642742731017):0.066043783465627):1.100383484076166):2.6340479153169807);")

label <- whalesnet$tip.label
library(ggimage)
p4 = ggevonet(whalesnet,layout = "slanted") +
    xlim(0,8) + ylim(0,15) +
    geom_tiplab(geom='label', offset=0.5) +
    geom_tiplab(aes(image=paste0('figures/',label, '.png')), geom="image",
                offset=0)
```
```{r whalesnet, out.width='85%', fig.cap="Explicit network for whales", fig.pos='H', fig.align='center'}
p4
```

# Discussion

The area of phylogenetic networks is very active. There are multiple developments on the inference of networks, as well as on the visualization. Furthermore, there are many new developments in the area of comparative methods on networks \missing{(references: bastide, brian)}. Our R package `tanggle` is certainly not the first software for the visualization of networks. Dendroscope \missing{(reference)}, and SplitsTree \missing{(reference)} are the pioneers of network visualization, with multiple functionalities being added constantly. In R, the widely used `ape` package \missing{(reference)} has extended the base plot function to visualize explicit networks, but no functionalities still to plot implicit networks. In julia, `PhyloPlots` is the accompanying package of `PhyloNetworks` \missing{(reference)} which is able to plot rooted explicit networks, but no implicit networks or unrooted.

`tanggle` is complementing the `ape` plotting function in two main ways: 1) allowing to plot implicit networks, 2) building on the flexibility of `ggtree` \missing{(reference)} (which in turns builds on the flexibility of `ggplot2` \missing{(reference)}) to manipulate the structure, nodes and edges in the network.
With the increase (and timely) pressure for reproducible research, limiting the use of design software like Adobe Illustrator to produce publication figures is paramount. The R package `ggtree` and now `tanggle` work towards script-based publication figures that can incorporate colors, images and mapped morphological and geographical characteristics.
