\documentclass[]{article}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\usepackage{fixltx2e} % provides \textsubscript
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
\else % if luatex or xelatex
  \ifxetex
    \usepackage{mathspec}
  \else
    \usepackage{fontspec}
  \fi
  \defaultfontfeatures{Ligatures=TeX,Scale=MatchLowercase}
\fi
% use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
% use microtype if available
\IfFileExists{microtype.sty}{%
\usepackage{microtype}
\UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\usepackage[margin=0.8in]{geometry}
\usepackage{hyperref}
\hypersetup{unicode=true,
            pdftitle={tanggle: An R package for the visualization of implicit and explicit phylogenetic networks},
            pdfauthor={klaus, marta, leann, francisco, eren, claudia},
            pdfborder={0 0 0},
            breaklinks=true}
\urlstyle{same}  % don't use monospace font for urls
\usepackage{natbib}
\bibliographystyle{apalike}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\usepackage{graphicx,grffile}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
\IfFileExists{parskip.sty}{%
\usepackage{parskip}
}{% else
\setlength{\parindent}{0pt}
\setlength{\parskip}{6pt plus 2pt minus 1pt}
}
\setlength{\emergencystretch}{3em}  % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{0}
% Redefines (sub)paragraphs to behave more like sections
\ifx\paragraph\undefined\else
\let\oldparagraph\paragraph
\renewcommand{\paragraph}[1]{\oldparagraph{#1}\mbox{}}
\fi
\ifx\subparagraph\undefined\else
\let\oldsubparagraph\subparagraph
\renewcommand{\subparagraph}[1]{\oldsubparagraph{#1}\mbox{}}
\fi

%%% Use protect on footnotes to avoid problems with footnotes in titles
\let\rmarkdownfootnote\footnote%
\def\footnote{\protect\rmarkdownfootnote}

%%% Change title format to be more compact
\usepackage{titling}

% Create subtitle command for use in maketitle
\providecommand{\subtitle}[1]{
  \posttitle{
    \begin{center}\large#1\end{center}
    }
}

\setlength{\droptitle}{-2em}

  \title{tanggle: An R package for the visualization of implicit and explicit
phylogenetic networks}
    \pretitle{\vspace{\droptitle}\centering\huge}
  \posttitle{\par}
    \author{klaus, marta, leann, francisco, eren, claudia}
    \preauthor{\centering\large\emph}
  \postauthor{\par}
    \date{}
    \predate{}\postdate{}
  
\usepackage[dvipsnames]{xcolor}
\newcommand{\missing}[1]{\textcolor{red}{\textbf{#1}}}
\newcommand{\revision}[1]{\textcolor{blue}{#1}}
\usepackage{setspace}

\doublespacing
\usepackage{array,ragged2e}
\usepackage{float}
\usepackage{lineno}
\linenumbers

\begin{document}
\maketitle

\begin{center}
\textbf{Abstract} 
\end{center}

Here we present an extension to the widely used visualization package
\texttt{ggtree} to the case of phylogenetic networks. Our packages
allows a variety of input data from DNA sequences to extended Newick
format, and allows the user to plot the networks with great flexibility
given that the many functionalities such as\ldots{}

\hypertarget{introduction}{%
\section{Introduction}\label{introduction}}

Phylogenetic trees are used to represent evolutionary relationships
between organisms. However, the tree is a strictly bifurcating structure
that does not capture reticulate evolution like horizontal gene
transfer, hybridization or introgression.

Recent years have seen an explosion of phylogenetic network inference
methods (\textcolor{red}{\textbf{cite everybody}}). These methods can be
divided into two main classes: methods to reconstruct implicit (or
split) networks, and methods to reconstruct explicit networks. Implicit
networks are data-displayed objects that include two (or more) options
for every uncertain split. For example, in figure
\textcolor{red}{\textbf{bla}}, there are two possible ways to resolve
the uncertain split, but the input data (DNA sequences or gene trees)
does not provide enough information to distinguish which one is the
``correct'' split. Or alternatively, both splits are ``correct'' in that
there is a pattern of gene flow among these species. Split networks are
fast to reconstruct, but they are limited in their interpretations are
they cannot tell us if the discordant splits are due to gene tree
estimation error, ILS or actual gene flow. Explicit networks, on the
contrary, assign a biological mechanism to every internal node, and thus
they are easier to interpret. For example, in figure
\textcolor{red}{\textbf{bla}}, the hybrid node corresponds to some
reticulation event, and the tree nodes correspond to speciation events.
Explicit networks are still limited in that they cannot identify the
specific biological mechanism behind a hybrid node (hybridization,
horizontal gene transfer, introgression), but hybrid edges are
parametrized by an inheritance probability (\(\gamma\)) that can provide
more information. For example, an estimated \(\hat{\gamma} \approx 0.5\)
could indicate a hybridization event, whereas smaller estimated
\(\hat{\gamma} \approx 0.01\) could refer to horizontal gene transfer.
The downside of explicit networks is that they are harder to
reconstruct, and existing inference
methods\textcolor{red}{\textbf{(reference)}} are not scalable enough to
handle big data.

\begin{figure}

{\centering \includegraphics[width=0.35\linewidth]{figures/explicit} \includegraphics[width=0.35\linewidth]{figures/implicit} 

}

\caption{\textbf{Left:} Explicit network. \textbf{Right:} Split network.}\label{fig:nettypes}
\end{figure}

\vspace{-0.25cm}

\hypertarget{extended-newick-format}{%
\subsection{Extended Newick format}\label{extended-newick-format}}

Traditionally, phylogenetic trees are written in Newick format
(\textcolor{red}{\textbf{reference}}). To represent network objects, we
use the extended Newick format (Morin and Moret 2006; Than et al.~2008;
Cardona et al.~2008a, 2008b). This format uses the concept of minor
hybrid edges (edges with \(\gamma < 0.5\)) and major hybrid edges
(\(\gamma > 0.5\)). By default, we detach the minor hybrid edge at each
hybrid node to write the extended Newick description of a network as we
would for a tree, with a repeated label, that of the hybrid node ('\#H1'
in figure \ref{fig:netNewick}). This description can include edge
information, formatted as :length:support:\(\gamma\).

\begin{figure}[H]

{\centering \includegraphics[width=0.35\linewidth]{figures/extendedNewick} \includegraphics[width=0.35\linewidth]{figures/extendedNewick2} 

}

\caption{Extended Newick format for networks: (((A,(B)\#H1),(C,\#H1)),D); The network is written as a tree with two nodes having the same label \#H1.}\label{fig:netNewick}
\end{figure}

For example, the parenthetical format of the network in figure
\ref{fig:netNewick} can include \(\gamma\) values:

\texttt{(((A,(B)\#H1:::0.8),(C,\#H1:::0.2)),D);},

which are written after three colons, because there is no information
about branch length nor support for the hybrid edges. Other internal
edges (tree edges) have information about branch lengths, which follow
just one colon. These tree edges do not have information about
\(\gamma\), because all tree edges have \(\gamma=1\).

\hypertarget{tanggle-package}{%
\section{\texorpdfstring{\texttt{tanggle}
package}{tanggle package}}\label{tanggle-package}}

The \texttt{tanggle} package extends the already widely used
\texttt{ggtree} (\textcolor{red}{\textbf{references}}) to implicit and
explicit phylogenetic networks. Our package has two main functions:
\texttt{ggsplitnet} that plots split networks, and \texttt{ggevonet}
that plots explicit networks.

\vspace{-0.25cm}

\hypertarget{reconstructing-and-plotting-a-split-network}{%
\subsection{Reconstructing and plotting a split
network}\label{reconstructing-and-plotting-a-split-network}}

There are three types of input data that our package takes to get a
split network:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Nexus file created with SplitsTree
  (\textcolor{red}{\textbf{reference}}) which is read with the
  \texttt{read.nexus.networx} function in the \texttt{phangorn} package.
\item
  Collection of gene trees (e.g.~from
  RAxML\textcolor{red}{\textbf{(reference)}} or
  MrBayes\textcolor{red}{\textbf{(reference)}}) in a nexus file, or in a
  text file with one row per gene tree in Newick format. These trees are
  read with the functions \texttt{read.tree} or \texttt{read.nexus}. A
  consensus split network is then computed using the
  \texttt{consensusNet} function in the \texttt{phangorn} package
  \textcolor{red}{\textbf{(reference)}}. This function implements the
  algorithm in \textcolor{red}{\textbf{(reference)}}.
\item
  Sequences in nexus, fasta or phylip format read with the
  \texttt{read.phyDat} function in \texttt{phangorn} package or the
  \texttt{read.dna} function in \texttt{ape} package. Distance matrices
  are computed then for specific models of evolution using the function
  \texttt{dist.ml} from \texttt{phangorn} package or the function
  \texttt{dist.dna} from \texttt{ape} package. From the distance matrix,
  a split network is reconstructed using the \texttt{neighborNet}
  function in the \texttt{phangorn} package. This function implements
  the algorithm in \textcolor{red}{\textbf{(reference)}}. In this case,
  there is the option to estimate branch lengths as well (which is not
  usually the case for split networks) with the function
  \texttt{splitsNetworks} in the \texttt{phangorn} package.
\end{enumerate}

After any of these three steps, we have a \texttt{networkx} object that
represents a split network by two matrices: 1) standard edge matrix with
two columns for the two nodes connecting every edge (row), and 2) split
vector with integer identifier for every edge. Each entry in this vector
represents the split that each edge is representing. In split networks,
there are more edges than splits as two (or more) edges could represent
the same split (see figure \textcolor{red}{\textbf{bla}}).

The \texttt{ggsplitnet} function takes a \texttt{networkx} object as
input, and produces a plot (see Figure \ref{fig:enet} left).

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(ggnetworx)}
\KeywordTok{data}\NormalTok{(yeast, }\DataTypeTok{package=}\StringTok{"phangorn"}\NormalTok{)}
\NormalTok{dm <-}\StringTok{ }\NormalTok{phangorn}\OperatorTok{::}\KeywordTok{dist.ml}\NormalTok{(yeast) }
\NormalTok{nnet <-}\StringTok{ }\NormalTok{phangorn}\OperatorTok{::}\KeywordTok{neighborNet}\NormalTok{(dm)}
\NormalTok{p =}\StringTok{ }\KeywordTok{ggsplitnet}\NormalTok{(nnet) }\OperatorTok{+}\StringTok{ }\KeywordTok{geom_tiplab2}\NormalTok{(}\DataTypeTok{cex=}\FloatTok{2.5}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{plotting-explicit-networks}{%
\subsection{Plotting explicit
networks}\label{plotting-explicit-networks}}

Algorithms to reconstruct explicit networks are more computationally
intensive than those to reconstruct split networks. Thus, to the best of
our knowledge, there are no R functions that would estimate an explicit
network from sequences or from gene trees. To obtain an explicit
network, users need to use existing tools like
PhyloNet\textcolor{red}{\textbf{(reference)}},
BEAST\textcolor{red}{\textbf{(reference)}} or
SNaQ\textcolor{red}{\textbf{(reference)}}. These methods take sequences
or gene trees as input and estimate an explicit network which is
represented in extended Newick format. The explicit network is read with
the \texttt{ape} function \texttt{read.evonet} and an \texttt{evonet}
object is created.

The \texttt{evonet} object has two matrices: 1) standard edge matrix
with two columns for the two nodes connecting every edge (row), and 2)
hybrid edges matrix with two columns for the nodes connected by the
hybrid edge (row). See Figure \ref{fig:enet} right for the plot.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(ape)}
\NormalTok{net =}\StringTok{ "((a:2,(b:1)#H1:1):1,(#H1,c:1):2);"}
\NormalTok{enet <-}\StringTok{ }\KeywordTok{read.evonet}\NormalTok{(}\DataTypeTok{text=}\NormalTok{net)}
\NormalTok{p2 =}\StringTok{ }\KeywordTok{ggevonet}\NormalTok{(enet) }\OperatorTok{+}\StringTok{ }\KeywordTok{geom_tiplab}\NormalTok{() }
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics[width=0.5\linewidth]{software-note_files/figure-latex/enet-1} 

}

\caption{\textbf{Left:} Split network for yeast data. \textbf{Right:} Explicit network from extended Newick format}\label{fig:enet}
\end{figure}

\hypertarget{split-network-for-the-neobatrachus-genus}{%
\section{\texorpdfstring{Split network for the \emph{Neobatrachus}
genus}{Split network for the Neobatrachus genus}}\label{split-network-for-the-neobatrachus-genus}}

\textcolor{red}{\textbf{(Marta: add here short description of genus)}}.

\begin{figure}[H]

{\centering \includegraphics[width=0.6\linewidth]{software-note_files/figure-latex/frognet-1} 

}

\caption{Explicit network for Neobatrachus genus frogs}\label{fig:frognet}
\end{figure}

\hypertarget{explicit-network-for-whales}{%
\section{Explicit network for
whales}\label{explicit-network-for-whales}}

Baleen whales (Mysticeti) represent the largest animals on Earth. Their
evolutionary history is complex given the lack of physical barriers in
the ocean, and the fact that they hybridize frequently. Arnason et al
\textcolor{red}{\textbf{(cite Arnason 2018)}} reconstruct the
network-like evolutionary history from a variety of methods, using
34,192 genome fragments, each with 20 kbp long. Reconstructed GF trees
with RAxML and GTR with gamma-distributed rate variation with invariable
sites. Here we use SNaQ \textcolor{red}{\textbf{(cite)}} to reconstruct
the network with two hybridization events, one of which was already
reported in the original paper (from blue whale to gray whale). SNaQ
produces the network in extended newick format, which we can read into R
and then plot with \texttt{tanggle}.

\begin{figure}[H]

{\centering \includegraphics[width=0.85\linewidth]{software-note_files/figure-latex/whalesnet-1} 

}

\caption{Explicit network for whales}\label{fig:whalesnet}
\end{figure}

\hypertarget{discussion}{%
\section{Discussion}\label{discussion}}

The area of phylogenetic networks is very active. There are multiple
developments on the inference of networks, as well as on the
visualization. Furthermore, there are many new developments in the area
of comparative methods on networks
\textcolor{red}{\textbf{(references: bastide, brian)}}. Our R package
\texttt{tanggle} is certainly not the first software for the
visualization of networks. Dendroscope
\textcolor{red}{\textbf{(reference)}}, and SplitsTree
\textcolor{red}{\textbf{(reference)}} are the pioneers of network
visualization, with multiple functionalities being added constantly. In
R, the widely used \texttt{ape} package
\textcolor{red}{\textbf{(reference)}} has extended the base plot
function to visualize explicit networks, but no functionalities still to
plot implicit networks. In julia, \texttt{PhyloPlots} is the
accompanying package of \texttt{PhyloNetworks}
\textcolor{red}{\textbf{(reference)}} which is able to plot rooted
explicit networks, but no implicit networks or unrooted.

\texttt{tanggle} is complementing the \texttt{ape} plotting function in
two main ways: 1) allowing to plot implicit networks, 2) building on the
flexibility of \texttt{ggtree} \textcolor{red}{\textbf{(reference)}}
(which in turns builds on the flexibility of \texttt{ggplot2}
\textcolor{red}{\textbf{(reference)}}) to manipulate the structure,
nodes and edges in the network. With the increase (and timely) pressure
for reproducible research, limiting the use of design software like
Adobe Illustrator to produce publication figures is paramount. The R
package \texttt{ggtree} and now \texttt{tanggle} work towards
script-based publication figures that can incorporate colors, images and
mapped morphological and geographical characteristics.

\bibliography{ggnetworx.bib}


\end{document}
